### PR TITLE
Service Fabric hosting

### PR DESCRIPTION
**Pull request by SeanFeldman**
Connects to https://github.com/Particular/PlatformDevelopment/issues/1112

SF hosting options (replaces PR #2594)

**TODO:**
- [x] Stateless service
- [x] Stateful service
- [x] Guest executable
- [x] Update [Azure index page](https://docs.particular.net/nservicebus/azure/) with self hosting option for SF
- [x] Hosting [NSB Gateway](https://github.com/Particular/NServiceBus.Gateway/pull/66)

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@danielmarbach 9 hours ago
>Could someone also have the question:
>Can I host NSB in an actor system and using NSB in the actors?
>I think it is a valid (but potentially dangerous) thought to have and we should address it

**Comment by SeanFeldman** [comment]
I've added a doco page + a snippet for hosting NSB with SF.
@tmasternak @yvesgoeleven Please have a look.

**Comment by SeanFeldman** [comment]
Hosting NSB Gateway feel alien to this page. Perhaps it should be a sub-page @yvesgoeleven ?

**Comment by SeanFeldman** [comment]
@danielmarbach that would probably not perform de-duplication that NSB gateway does. Aside from that, no idea. @yvesgoeleven, did you have a specific reason to suggest hosting of Gateway with SF in mind?

PS: @danielmarbach your router should have cut inet access at this hour 😛

**Comment by yvesgoeleven** [comment]
@SeanFeldman I'll take care of the gateway section

File: Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs [hunk 1]
```diff
@@ -1,24 +1,37 @@
...
+    #region StatelessEndpointCommunicationListener
 
-    public class EndpointCommunicationListener : ICommunicationListener
+    public class StatelessEndpointCommunicationListener : ICommunicationListener
     {
         IEndpointInstance endpointInstance;
 
+        public StatelessEndpointCommunicationListener(IReliableStateManager stateManager)
+        {
+            throw new System.NotImplementedException();
```
Comment (danielmarbach) [inline Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs]: Hmm? why does this throw

File: Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs [hunk 2]
```diff
@@ -37,4 +50,79 @@ public void Abort()
...
+    public class StatefulEndpointCommunicationListener : ICommunicationListener
+    {
+        readonly IReliableStateManager stateManager;
+        IEndpointInstance endpointInstance;
+        EndpointConfiguration endpointConfiguration;
+
+        public StatefulEndpointCommunicationListener(IReliableStateManager stateManager)
+        {
+            this.stateManager = stateManager;
+            throw new System.NotImplementedException();
```
Comment (danielmarbach) [inline Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs]: Hmmm? Why does this throw?

File: Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs [hunk 3]
```diff
@@ -37,4 +50,79 @@ public void Abort()
...
+        {
+            new ServiceReplicaListener(context => listener)
+        };
+        }
+
+        protected override async Task RunAsync(CancellationToken cancellationToken)
+        {
+            await listener.RunAsync();
+
+            await base.RunAsync(cancellationToken);
```
Comment (danielmarbach) [inline Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs]: If I recall it is not required to run base, see impl

```
protected virtual Task RunAsync(CancellationToken cancellationToken)
    {
      return (Task) Task.FromResult<bool>(true);
    }
```

File: Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs (hunk 1 above)
Comment (yvesgoeleven) [inline Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs]: because I'm an idiot and will fix it

**Comment by SeanFeldman** [comment]
@yvesgoeleven added a few questions. Also pinged Adam to ensure we didn't miss anything.

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+---
+title: Hosting Gateway with Service Fabric
+related:
+ - nservicebus/service-fabric
+reviewed: 2017-03-30
+---
+
+When adopting Service Fabric, it's not uncommon that the Service Fabric hosted endpoints need to interact with endpoints outside of the cluster. This can get tricky especially when the endpoints inside Service Fabric are stateful. When integrating using client side distribution, or when using the Service Fabric built in [reverse proxy](https://docs.microsoft.com/nl-nl/azure/service-fabric/service-fabric-reverseproxy) to expose the endpoints as web services, then the partition information needs to be provided by the consumer, which is often not desired.
```
Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Do we need it here? Can we just assume that the user wants to use `Gateway` and knows what are the valid reasons based on `Gateway` [documentation](https://docs.particular.net/nservicebus/gateway/#logically-different-sites)

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+Alternatively the [nservicebus gateway](/nservicebus/gateway/) can be leveraged as an intermediary to solve this problem. It provides reliable request reply semantics, with deduplication, between sites.
+
+To host NServiceBus Gateway in an endpoint deployed to Service Fabric, the following has to be taken into account:
+
+1. Host the gateway as a [stateless service](/nservicebus/hosting/service-fabric/hosting/#stateful service), and use [partition aware routing](/samples/azure/azure-service-fabric-routing/) within to forward messages to other parts of the cluster.
+2. Make sure the endpoint is present on all service fabric cluster instances, by specifying -1 for the `InstanceCount` value.
+3. Add a second communication listener that opens up communication on the gateway address using the local FQDN
+
+(Similar to https://github.com/adam3039/SampleNsbAsfProject/blob/master/WebApi/OwinCommunicationListener.cs#L95, just return the right address but don't open a port as the gateway is already doing that)
```
Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Should we reference external repository? I think we should just add snippet.

Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: I was looking at the sample and I think we need to extend it a bit if we want to use parts as snippets.

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+reviewed: 2017-03-30
+---
+
+When adopting Service Fabric, it's not uncommon that the Service Fabric hosted endpoints need to interact with endpoints outside of the cluster. This can get tricky especially when the endpoints inside Service Fabric are stateful. When integrating using client side distribution, or when using the Service Fabric built in [reverse proxy](https://docs.microsoft.com/nl-nl/azure/service-fabric/service-fabric-reverseproxy) to expose the endpoints as web services, then the partition information needs to be provided by the consumer, which is often not desired.
+
+Alternatively the [nservicebus gateway](/nservicebus/gateway/) can be leveraged as an intermediary to solve this problem. It provides reliable request reply semantics, with deduplication, between sites.
+
+To host NServiceBus Gateway in an endpoint deployed to Service Fabric, the following has to be taken into account:
+
+1. Host the gateway as a [stateless service](/nservicebus/hosting/service-fabric/hosting/#stateful service), and use [partition aware routing](/samples/azure/azure-service-fabric-routing/) within to forward messages to other parts of the cluster.
```
Comment (yvesgoeleven) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: stateless services are easy to expose, just add a communication listerer, no need for a gateway

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md (hunk 4 above)
Comment (yvesgoeleven) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Isn't this an addition to the already present reasons? Do not want to leak out the partitioning even if it is part of the same logical site...

Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: I'm on the fence here. I'm not sure we want to suggest using `Gateway` as a substitute for Client-Side Routing. I think we should tell users to `Gateway` for the purposes it was designed - which I think are still valid for FS

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+2. Because the service instance are hosted behind a load balancer, it is required to configure the gateway channel address to use a URL with wildcard. The reason is that the public IP address is the one from the cluster load balancer and will therefore differ from the one that the gateway communication listener will effectively be listening on. 
+
+Snippet: configureWildcardGatewayChannel
+
+3. Open the correct port for the gateway on the cluster load balancer by specifying an `Endpoint` entry in the service manifest.
+
+Snippet: serviceManifestEndpoint
+
+4. Add a second communication listener that returns the gateway address using the local FQDN instead of the + sign so that the service fabric instance is aware of the local address.
```
Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Do we want to mention that one of the reasons is proper registration of `Gateway` in `NamingService`?

Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: It's for the ACL. I don't think it's worth mentioning it here.

**Comment by tmasternak** [comment]
@SeanFeldman it's ready for review. After that we can merge.

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md (hunk 4 above)
Comment (SeanFeldman) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: The link is pointing to "nl-nl". Should it point to "en-us" since we target that locale?

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md (hunk 7 above)
Comment (SeanFeldman) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: @tmasternak Like in SF NamingService? Gateway doesn't do that. It's just for the ACL.

File: Snippets/Gateway/Gateway_2/ServiceFabric.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    using Microsoft.ServiceFabric.Services.Communication.Runtime;
+
+    #region GatewayCommunicationListener
+    public class GatewayCommunicationListener : ICommunicationListener
+    {
+        readonly StatelessServiceContext _serviceContext;
+
+        public GatewayCommunicationListener(StatelessServiceContext serviceContext)
+        {
+            _serviceContext = serviceContext;
```
Comment (SeanFeldman) [inline Snippets/Gateway/Gateway_2/ServiceFabric.cs]: Change `_serviceContext` to `serviceContext` to be consistent.

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md (hunk 7 above)
Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: For me this is a confusing. AFAIK when a `ICommunicationListener` returns a url SF does at least two things:
 * It configures `urlacl` for that url for process running the service - this is also done by by `Gateway` itself
 * It registers the url in `NamingService` so that it's discoverable inside the cluster - this is not needed in the approach we guide users to use - `Gateway` behind Load Balancer and deployed on all the nodes.

In other words for me it's not clear what `service fabric instance is aware of the local address` means and secondly if this is even needed.

Comment (yvesgoeleven) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: I'll reword it

Comment (yvesgoeleven) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: btw, I don't think gateway can create the ACL, services do not have administrative priviliges to do so

**Comment by SeanFeldman** [comment]
With 1:3 votes, majority voted to keep the gateway info with a warning about databus.
image

File: Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            listener = new StatefulEndpointCommunicationListener(StateManager);
+            return new List<ServiceReplicaListener>
+            {
+                new ServiceReplicaListener(context => listener)
+            };
+        }
+
+        protected override async Task RunAsync(CancellationToken cancellationToken)
+        {
+            await listener.RunAsync();
```
Comment (danielmarbach) [inline Snippets/Core/Core_6/ServiceFabricHosting/EndpointCommunicationListener.cs]: Should we return instead of await?

File: Snippets/Gateway/Gateway_2/project.json [hunk 10]
```diff
@@ -1,5 +1,6 @@
 ﻿{
   "dependencies": {
+    "Microsoft.ServiceFabric.Services": "2.5.216",
```
Comment (danielmarbach) [inline Snippets/Gateway/Gateway_2/project.json]: https://github.com/Particular/docs.particular.net/pull/2602/files#diff-290f49db3155932839f03b1f8c0b0efa uses `2.*`. Should we align?

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
+---
+title: Hosting Gateway with Service Fabric
+related:
+ - nservicebus/service-fabric
+reviewed: 2017-03-30
+---
+
+WARNING: [Gateway](/nservicebus/gateway) hosted in Service Fabric does not support forwarding messages containing [Databus](/nservicebus/messaging/databus/) properties.
+
+When adopting Service Fabric, it's not uncommon that the Service Fabric hosted endpoints need to interact with endpoints outside of the cluster. This can get tricky especially when the endpoints inside Service Fabric are stateful. When integrating using client side distribution, or when using the Service Fabric built in [reverse proxy](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy) to expose the endpoints as web services, then the partition information needs to be provided by the consumer, which is often not desired.
```
Comment (danielmarbach) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Should the client side distribution be renamed to sender side distribution?

Also this argument here seems to be a bit strange for me since we have a partition affine routing sample which shows how to deal with clients outside the cluster scenario. Should we reword a bit and link to the routing sample? For me this reads like we are actively encouraging the gateway instead the routing sample.

File: nservicebus/hosting/service-fabric-hosting/index.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+1. Stateful services
+1. Guest executable 
+
+To decide what option to use, refer to [Service Fabric documentation](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-overview).
+
+Note: The Actor model is another Service Fabric programming model that is currently not supported by NServiceBus.
+
+### Stateless service
+
+Hosting with a Stateless service is very similar to any other Azure-based hosting (using [Cloud services](/nservicebus/hosting/cloud-services-host) or [self-hosting](/nservicebus/hosting/#self-hosting) with [Azure App Service](https://docs.microsoft.com/en-us/azure/app-service/)). Endpoints are stateless and use storage external to Service Fabric for managing data needed for their operation. Endpoints can be scaled out, leveraging competing consumer at the transport level.
```
Comment (danielmarbach) [inline nservicebus/hosting/service-fabric-hosting/index.md]: Could we here link to scaleout doco where we mention competing consumers?

File: nservicebus/hosting/service-fabric-hosting/index.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+snippet: StatefulEndpointCommunicationListener
+
+snippet: StatefulService
+
+Due to characteristics of Stateful Services i.e. data partitioning and local data storage, additional configuration aspects have to be taken into consideration: 
+
+- Service partitioning schema must be well defined before the first deployment occurs (repartitioning requires all data to be deleted)
+- Messages must be routed among the shards according to the partitioning schema
+- There can only be one NServiceBus endpoint running per service partition (in the master instance)
```
Comment (danielmarbach) [inline nservicebus/hosting/service-fabric-hosting/index.md]: Can you explain this? This sounds like it is impossible to run multiple endpoints on the master instance which I don't see why it would not be possible

Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/index.md]: I agree it's not very clear as it is state right now. How about: "Endpoint instances can be running only on primary replicas"?

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md (hunk 11 above)
Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: @danielmarbach I've raised more or less the same question in https://github.com/Particular/docs.particular.net/pull/2602/files/a3e110688ceeee48f54400f0e6893b5e08f50285#diff-6b1e8e331b24367abdc1a0fdee54d382. But we decided to leave it as is.

Comment (tmasternak) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Changed `client side distribution` to `sender side distribution`

File: nservicebus/hosting/service-fabric-hosting/index.md (hunk 13 above)
Comment (SeanFeldman) [inline nservicebus/hosting/service-fabric-hosting/index.md]: Personally, I find "Endpoint instances can be running only on primary replicas" confusing as well. Bottom line that we want to say is that "there will be only one active instance of an endpoint per partition".

Comment (SeanFeldman) [inline nservicebus/hosting/service-fabric-hosting/index.md]: In regards to https://github.com/Particular/docs.particular.net/pull/2602#discussion_r109865600 - not impossible, but backwards. 

@danielmarbach what would be the point of running multiple instance of the same logical endpoint (multi-hosting) on the same primary replica for a given partition? It's resource wasteful. Rather dial the concurrency up. Perhaps we need to discuss this in a synchronous way.

File: nservicebus/hosting/service-fabric-hosting/hosting-gateway.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
+---
+title: Hosting Gateway with Service Fabric
+related:
+ - nservicebus/service-fabric
+reviewed: 2017-03-30
+---
+
+WARNING: [Gateway](/nservicebus/gateway) hosted in Service Fabric does not support forwarding messages containing [Databus](/nservicebus/messaging/databus/) properties.
+
+When adopting Service Fabric, it's not uncommon that the Service Fabric hosted endpoints need to interact with endpoints outside of the cluster. This can get tricky especially when the endpoints inside Service Fabric are stateful. When integrating using sender side distribution, or when using the Service Fabric built in [reverse proxy](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reverseproxy) to expose the endpoints as web services, then the partition information needs to be provided by the consumer, which is often not desired.
```
Comment (SeanFeldman) [inline nservicebus/hosting/service-fabric-hosting/hosting-gateway.md]: Going to remove the 
> , which is often not desired

We shouldn't prefer gateway over custom connectivity or the other way arround.

**Comment by danielmarbach** [comment]
Will merge when green

**Comment by SeanFeldman** [comment]
Merged. Now I can sleep 😄
